certbot_dns_cpanel/installer.py

```diff
@@ -0,0 +1,187 @@
+"""certbot-dns-cpanel installer plugin"""
+
+import logging
+import json
+from configobj import ConfigObj
+
+try:
+    # python 3
+    from urllib.request import urlopen, Request
+    from urllib.parse import urlencode
+except ImportError:
+    # python 2
+    from urllib import urlencode
+    from urllib2 import urlopen, Request
+
+import zope.interface
+
+from certbot import errors
+from certbot import interfaces
+from certbot.plugins import common
+from certbot.plugins import dns_common
+
+logger = logging.getLogger(__name__)
+
+@zope.interface.implementer(interfaces.IInstaller)
+@zope.interface.provider(interfaces.IPluginFactory)
+class Installer(common.Plugin):
+
+    description = "Upload generated certificate in cPanel"
+    
+    def __init__(self, *args, **kwargs):
+        super(Installer, self).__init__(*args, **kwargs)
+
+    @classmethod
+    def add_parser_arguments(cls, add):
+        super(Installer, cls).add_parser_arguments(add)
+        add("credentials", help="The cPanel credentials INI file")
+
+    def prepare(self):  # pylint: disable=missing-docstring,no-self-use
+        pass  # pragma: no cover
+
+    def more_info(self):  # pylint: disable=missing-docstring,no-self-use
+        return self.description
+
+    def get_all_names(self):  # pylint: disable=missing-docstring,no-self-use
+        pass  # pragma: no cover
+
+    def _validate_credentials(self, credentials):
+        url = credentials.conf('url')
+        username  = credentials.conf('username')
+        token = credentials.conf('token')
+        password = credentials.conf('password')
+
+        if not url:
+            raise errors.PluginError('%s: url is required' % credentials.confobj.filename)
+
+        if not username:
+            raise errors.PluginError('%s: username and token (prefered) or password are required' % credentials.confobj.filename)
+
+        if token:
+            if password:
+                logger.warning('%s: token and password are exclusive, token will be used when both are provided' % credentials.confobj.filename)
+        elif not password:
+            raise errors.PluginError('%s: password or token (prefered) are required' % credentials.confobj.filename)
+
+    def _get_cpanel_client(self):
+        logger.debug("Credentials file: '%s'", self.conf('credentials'))
+        dnsauth = dns_common.DNSAuthenticator(self.config, __name__)
```

Comment:
Creating a DNSAuthenticator just to parse the config file is kind of ugly.

I think the way to go would be to merge the 2 classes and have the authenticator and installer plugin performed by the same class. This is even suggested in the certbot documentation: https://certbot.eff.org/docs/contributing.html#installer
> Installers and Authenticators will oftentimes be the same class/object (because for instance both tasks can be performed by a webserver like nginx) though this is not always the case (the standalone plugin is an authenticator that listens on port 80, but it cannot install certificates; a postfix plugin would be an installer but not an authenticator).

This will allow reuse of all the configuration parsing/cpanel api object and keep us from repeating ourselves.